GracefulStop() is racy, leading to double-closing a channel

Observed in CockroachDB.

I had to instrument grpc as follows:
```diff
diff --git a/transport/http2_server.go b/transport/http2_server.go
index 2322c93..1502942 100644
--- a/transport/http2_server.go
+++ b/transport/http2_server.go
@@ -37,11 +37,14 @@ import (
 	"bytes"
 	"errors"
 	"io"
+	"log"
 	"math"
 	"net"
 	"strconv"
 	"sync"
 
+	"github.com/cockroachdb/cockroach/util/caller"
+
 	"golang.org/x/net/context"
 	"golang.org/x/net/http2"
 	"golang.org/x/net/http2/hpack"
@@ -56,6 +59,12 @@ import (
 // the stream's state.
 var ErrIllegalHeaderWrite = errors.New("transport: the stream is done or WriteHeader was already called")
 
+type closer struct {
+	file string
+	line int
+	fun  string
+}
+
 // http2Server implements the ServerTransport interface with HTTP2.
 type http2Server struct {
 	conn        net.Conn
@@ -73,6 +82,8 @@ type http2Server struct {
 	hBuf         *bytes.Buffer  // the buffer for HPACK encoding
 	hEnc         *hpack.Encoder // HPACK encoder
 
+	closers []closer
+
 	// The max number of concurrent streams.
 	maxStreams uint32
 	// controlBuf delivers all the control related tasks (e.g., window
@@ -265,6 +276,7 @@ func (t *http2Server) HandleStreams(handle func(*Stream)) {
 				t.controlBuf.put(&resetStream{se.StreamID, se.Code})
 				continue
 			}
+			grpclog.Printf("transport: http2Server.HandleStreams failed to read frame: %v", err)
 			t.Close()
 			return
 		}
@@ -711,6 +723,15 @@ func (t *http2Server) Close() (err error) {
 		t.mu.Unlock()
 		return errors.New("transport: Close() was already called")
 	}
+	hadCloser := len(t.closers) > 0
+	file, line, fun := caller.Lookup(1)
+	t.closers = append(t.closers, closer{file, line, fun})
+	if hadCloser {
+		log.Printf("MULTIPLE CLOSERS IN STATE %v", t.state)
+		for _, closer := range t.closers {
+			log.Printf("%s:%d (%s)", closer.file, closer.line, closer.fun)
+		}
+	}
 	t.state = closing
 	streams := t.activeStreams
 	t.activeStreams = nil
```

Then running some tests that actually use GracefulStop() I see:
```
I160727 16:11:53.281672 stopper.go:408  quiesceing; tasks left: // <--- this is where GracefulStop() is called.
1      storage/queue.go:427
I160727 16:11:53.281679 http2_server.go:279  transport: http2Server.HandleStreams failed to read frame: EOF
I160727 16:11:53.281754 http2_server.go:730  MULTIPLE CLOSERS IN STATE 3
I160727 16:11:53.281763 http2_server.go:732  http2_server.go:280 ((*http2Server).HandleStreams)
I160727 16:11:53.281768 http2_server.go:732  /Users/tamird/src/go/src/google.golang.org/grpc/server.go:411 ((*Server).serveStreams)
panic: close of closed channel

goroutine 5194 [running]:
panic(0x50265a0, 0xc8200da9c0)
	/Users/tamird/src/go1.6/src/runtime/panic.go:481 +0x3e6
google.golang.org/grpc/transport.(*http2Server).Close(0xc8200f04d0, 0x0, 0x0)
	/Users/tamird/src/go/src/google.golang.org/grpc/transport/http2_server.go:739 +0x61f
google.golang.org/grpc.(*Server).serveStreams(0xc8205899a0, 0x96e1648, 0xc8200f04d0)
	/Users/tamird/src/go/src/google.golang.org/grpc/server.go:411 +0x168
google.golang.org/grpc.(*Server).serveNewHTTP2Transport(0xc8205899a0, 0x96e1548, 0xc820460300, 0x96e15a8, 0xc8200f0420)
	/Users/tamird/src/go/src/google.golang.org/grpc/server.go:396 +0x49d
google.golang.org/grpc.(*Server).handleRawConn(0xc8205899a0, 0x95589d8, 0xc8203ce0b0)
	/Users/tamird/src/go/src/google.golang.org/grpc/server.go:373 +0x4ee
created by google.golang.org/grpc.(*Server).Serve
	/Users/tamird/src/go/src/google.golang.org/grpc/server.go:345 +0x38c
```

So it seems that:
  1. the client hangs up, causing the EOF and `Close()`ing the stream and `close()`ing its channel
  1. GracefulStop() is called
  1. The run-loop in `(*http2Server).controller()` receives the `goAway` and sets the stream to `draining` again (which is illegal, since it is already closed)
  1. GracefulStop() `Close()`es the stream again, causing a double close on the channel and panicking.

--- a/clientconn.go
+++ b/clientconn.go
@@ -296,6 +296,8 @@ const (
 	TransientFailure
 	// Shutdown indicates the ClientConn has started shutting down.
 	Shutdown
+	// Drain
+	Drain
 )
 
 func (s ConnectivityState) String() string {
@@ -310,6 +312,8 @@ func (s ConnectivityState) String() string {
 		return "TRANSIENT_FAILURE"
 	case Shutdown:
 		return "SHUTDOWN"
+	case Drain:
+		return "DRAIN"
 	default:
 		panic(fmt.Sprintf("unknown connectivity state: %d", s))
 	}
@@ -632,7 +636,7 @@ func (ac *addrConn) transportMonitor() {
 		case <-t.Error():
 			ac.mu.Lock()
 			if ac.state == Shutdown {
-				// ac.tearDown(...) has been invoked.
+				// ac has been shutdown.
 				ac.mu.Unlock()
 				return
 			}

--- a/clientconn.go
+++ b/clientconn.go
@@ -624,12 +624,32 @@ func (ac *addrConn) transportMonitor() {
 		// shutdownChan is needed to detect the teardown when
 		// the addrConn is idle (i.e., no RPC in flight).
 		case <-ac.shutdownChan:
+			select {
+			case <-t.Error():
+				t.Close()
+			default:
+			}
 			return
 		case <-t.GoAway():
+			select {
+			case <-t.Error():
+				t.Close()
+				return
+			default:
+			}
 			ac.tearDown(errConnDrain)
 			ac.cc.newAddrConn(ac.addr, true)
 			return
 		case <-t.Error():
+			select {
+			case <-ac.shutdownChan:
+				t.Close()
+				return
+			case <-t.GoAway():
+				t.Close()
+				return
+			default:
+			}
 			ac.mu.Lock()
 			if ac.state == Shutdown {
 				// ac has been shutdown.

--- a/clientconn.go
+++ b/clientconn.go
@@ -296,8 +296,6 @@ const (
 	TransientFailure
 	// Shutdown indicates the ClientConn has started shutting down.
 	Shutdown
-	// Drain
-	Drain
 )
 
 func (s ConnectivityState) String() string {
@@ -312,8 +310,6 @@ func (s ConnectivityState) String() string {
 		return "TRANSIENT_FAILURE"
 	case Shutdown:
 		return "SHUTDOWN"
-	case Drain:
-		return "DRAIN"
 	default:
 		panic(fmt.Sprintf("unknown connectivity state: %d", s))
 	}

--- a/clientconn.go
+++ b/clientconn.go
@@ -68,7 +68,7 @@ var (
 	// errCredentialsConflict indicates that grpc.WithTransportCredentials()
 	// and grpc.WithInsecure() are both called for a connection.
 	errCredentialsConflict = errors.New("grpc: transport credentials are set for an insecure connection (grpc.WithTransportCredentials() and grpc.WithInsecure() are both called)")
-	// errNetworkIP indicates that the connection is down due to some network I/O error.
+	// errNetworkIO indicates that the connection is down due to some network I/O error.
 	errNetworkIO = errors.New("grpc: failed with network I/O error")
 	// errConnDrain indicates that the connection starts to be drained and does not accept any new RPCs.
 	errConnDrain = errors.New("grpc: the connection is drained")
@@ -631,13 +631,18 @@ func (ac *addrConn) transportMonitor() {
 			}
 			return
 		case <-t.GoAway():
+			// If GoAway happens without any network I/O error, ac is closed without shutting down the
+			// underlying transport (the transport will be closed when all the pending RPCs finished or
+			// failed.).
+			// If GoAway and some network I/O error happen concurrently, ac and its underlying transport
+			// are closed.
+			// In both cases, a new ac is created.
 			select {
 			case <-t.Error():
-				t.Close()
-				return
+				ac.tearDown(errNetworkIO)
 			default:
+				ac.tearDown(errConnDrain)
 			}
-			ac.tearDown(errConnDrain)
 			ac.cc.newAddrConn(ac.addr, true)
 			return
 		case <-t.Error():
@@ -646,7 +651,8 @@ func (ac *addrConn) transportMonitor() {
 				t.Close()
 				return
 			case <-t.GoAway():
-				t.Close()
+				ac.tearDown(errNetworkIO)
+				ac.cc.newAddrConn(ac.addr, true)
 				return
 			default:
 			}

--- a/server.go
+++ b/server.go
@@ -774,6 +774,8 @@ func (s *Server) Stop() {
 	s.lis = nil
 	st := s.conns
 	s.conns = nil
+	// interrupt GracefulStop if Stop and GracefulStop are called concurrently.
+	s.cv.Signal()
 	s.mu.Unlock()
 
 	for lis := range listeners {
@@ -803,13 +805,13 @@ func (s *Server) GracefulStop() {
 	for lis := range s.lis {
 		lis.Close()
 	}
+	s.lis = nil
 	for c := range s.conns {
 		c.(transport.ServerTransport).Drain()
 	}
 	for len(s.conns) != 0 {
 		s.cv.Wait()
 	}
-	s.lis = nil
 	s.conns = nil
 	if s.events != nil {
 		s.events.Finish()

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -2282,8 +2282,8 @@ func leakCheck(t testing.TB) func() {
 	}
 	return func() {
 		// Loop, waiting for goroutines to shut down.
-		// Wait up to 5 seconds, but finish as quickly as possible.
-		deadline := time.Now().Add(5 * time.Second)
+		// Wait up to 10 seconds, but finish as quickly as possible.
+		deadline := time.Now().Add(10 * time.Second)
 		for {
 			var leaked []string
 			for _, g := range interestingGoroutines() {

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -710,50 +710,19 @@ func testConcurrentClientConnCloseAndServerGoAway(t *testing.T, e env) {
 
 	cc := te.clientConn()
 	tc := testpb.NewTestServiceClient(cc)
-	stream, err := tc.FullDuplexCall(context.Background(), grpc.FailFast(false))
-	if err != nil {
-		t.Fatalf("%v.FullDuplexCall(_) = _, %v, want <nil>", tc, err)
-	}
-	// Finish an RPC to make sure the connection is good.
 	if _, err := tc.EmptyCall(context.Background(), &testpb.Empty{}, grpc.FailFast(false)); err != nil {
 		t.Fatalf("%v.EmptyCall(_, _, _) = _, %v, want _, <nil>", tc, err)
 	}
 	ch := make(chan struct{})
+	// Close ClientConn and Server concurrently.
 	go func() {
 		te.srv.GracefulStop()
 		close(ch)
 	}()
-	// Loop until the server side GoAway signal is propagated to the client.
-	for {
-		ctx, _ := context.WithTimeout(context.Background(), 10*time.Millisecond)
-		if _, err := tc.EmptyCall(ctx, &testpb.Empty{}, grpc.FailFast(false)); err == nil {
-			continue
-		}
-		break
-	}
-	// Stop the server and close all the connections.
-	te.srv.Stop()
-	respParam := []*testpb.ResponseParameters{
-		{
-			Size: proto.Int32(1),
-		},
-	}
-	payload, err := newPayload(testpb.PayloadType_COMPRESSABLE, int32(100))
-	if err != nil {
-		t.Fatal(err)
-	}
-	req := &testpb.StreamingOutputCallRequest{
-		ResponseType:       testpb.PayloadType_COMPRESSABLE.Enum(),
-		ResponseParameters: respParam,
-		Payload:            payload,
-	}
-	if err := stream.Send(req); err == nil {
-		if _, err := stream.Recv(); err == nil {
-			t.Fatalf("%v.Recv() = _, %v, want _, <nil>", stream, err)
-		}
-	}
+	go func() {
+		cc.Close()
+	}()
 	<-ch
-	awaitNewConnLogOutput()
 }
 
 func TestConcurrentServerStopAndGoAway(t *testing.T) {

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -686,6 +686,146 @@ func testServerGoAwayPendingRPC(t *testing.T, e env) {
 	awaitNewConnLogOutput()
 }
 
+func TestConcurrentClientConnCloseAndServerGoAway(t *testing.T) {
+	defer leakCheck(t)()
+	for _, e := range listTestEnv() {
+		if e.name == "handler-tls" {
+			continue
+		}
+		testConcurrentClientConnCloseAndServerGoAway(t, e)
+	}
+}
+
+func testConcurrentClientConnCloseAndServerGoAway(t *testing.T, e env) {
+	te := newTest(t, e)
+	te.userAgent = testAppUA
+	te.declareLogNoise(
+		"transport: http2Client.notifyError got notified that the client transport was broken EOF",
+		"grpc: Conn.transportMonitor exits due to: grpc: the client connection is closing",
+		"grpc: Conn.resetTransport failed to create client transport: connection error",
+		"grpc: Conn.resetTransport failed to create client transport: connection error: desc = \"transport: dial unix",
+	)
+	te.startServer(&testServer{security: e.security})
+	defer te.tearDown()
+
+	cc := te.clientConn()
+	tc := testpb.NewTestServiceClient(cc)
+	stream, err := tc.FullDuplexCall(context.Background(), grpc.FailFast(false))
+	if err != nil {
+		t.Fatalf("%v.FullDuplexCall(_) = _, %v, want <nil>", tc, err)
+	}
+	// Finish an RPC to make sure the connection is good.
+	if _, err := tc.EmptyCall(context.Background(), &testpb.Empty{}, grpc.FailFast(false)); err != nil {
+		t.Fatalf("%v.EmptyCall(_, _, _) = _, %v, want _, <nil>", tc, err)
+	}
+	ch := make(chan struct{})
+	go func() {
+		te.srv.GracefulStop()
+		close(ch)
+	}()
+	// Loop until the server side GoAway signal is propagated to the client.
+	for {
+		ctx, _ := context.WithTimeout(context.Background(), 10*time.Millisecond)
+		if _, err := tc.EmptyCall(ctx, &testpb.Empty{}, grpc.FailFast(false)); err == nil {
+			continue
+		}
+		break
+	}
+	// Stop the server and close all the connections.
+	te.srv.Stop()
+	respParam := []*testpb.ResponseParameters{
+		{
+			Size: proto.Int32(1),
+		},
+	}
+	payload, err := newPayload(testpb.PayloadType_COMPRESSABLE, int32(100))
+	if err != nil {
+		t.Fatal(err)
+	}
+	req := &testpb.StreamingOutputCallRequest{
+		ResponseType:       testpb.PayloadType_COMPRESSABLE.Enum(),
+		ResponseParameters: respParam,
+		Payload:            payload,
+	}
+	if err := stream.Send(req); err == nil {
+		if _, err := stream.Recv(); err == nil {
+			t.Fatalf("%v.Recv() = _, %v, want _, <nil>", stream, err)
+		}
+	}
+	<-ch
+	awaitNewConnLogOutput()
+}
+
+func TestConcurrentServerStopAndGoAway(t *testing.T) {
+	defer leakCheck(t)()
+	for _, e := range listTestEnv() {
+		if e.name == "handler-tls" {
+			continue
+		}
+		testConcurrentServerStopAndGoAway(t, e)
+	}
+}
+
+func testConcurrentServerStopAndGoAway(t *testing.T, e env) {
+	te := newTest(t, e)
+	te.userAgent = testAppUA
+	te.declareLogNoise(
+		"transport: http2Client.notifyError got notified that the client transport was broken EOF",
+		"grpc: Conn.transportMonitor exits due to: grpc: the client connection is closing",
+		"grpc: Conn.resetTransport failed to create client transport: connection error",
+		"grpc: Conn.resetTransport failed to create client transport: connection error: desc = \"transport: dial unix",
+	)
+	te.startServer(&testServer{security: e.security})
+	defer te.tearDown()
+
+	cc := te.clientConn()
+	tc := testpb.NewTestServiceClient(cc)
+	stream, err := tc.FullDuplexCall(context.Background(), grpc.FailFast(false))
+	if err != nil {
+		t.Fatalf("%v.FullDuplexCall(_) = _, %v, want <nil>", tc, err)
+	}
+	// Finish an RPC to make sure the connection is good.
+	if _, err := tc.EmptyCall(context.Background(), &testpb.Empty{}, grpc.FailFast(false)); err != nil {
+		t.Fatalf("%v.EmptyCall(_, _, _) = _, %v, want _, <nil>", tc, err)
+	}
+	ch := make(chan struct{})
+	go func() {
+		te.srv.GracefulStop()
+		close(ch)
+	}()
+	// Loop until the server side GoAway signal is propagated to the client.
+	for {
+		ctx, _ := context.WithTimeout(context.Background(), 10*time.Millisecond)
+		if _, err := tc.EmptyCall(ctx, &testpb.Empty{}, grpc.FailFast(false)); err == nil {
+			continue
+		}
+		break
+	}
+	// Stop the server and close all the connections.
+	te.srv.Stop()
+	respParam := []*testpb.ResponseParameters{
+		{
+			Size: proto.Int32(1),
+		},
+	}
+	payload, err := newPayload(testpb.PayloadType_COMPRESSABLE, int32(100))
+	if err != nil {
+		t.Fatal(err)
+	}
+	req := &testpb.StreamingOutputCallRequest{
+		ResponseType:       testpb.PayloadType_COMPRESSABLE.Enum(),
+		ResponseParameters: respParam,
+		Payload:            payload,
+	}
+	if err := stream.Send(req); err == nil {
+		if _, err := stream.Recv(); err == nil {
+			t.Fatalf("%v.Recv() = _, %v, want _, <nil>", stream, err)
+		}
+	}
+	<-ch
+	awaitNewConnLogOutput()
+}
+
 func TestFailFast(t *testing.T) {
 	defer leakCheck(t)()
 	for _, e := range listTestEnv() {

--- a/transport/http2_client.go
+++ b/transport/http2_client.go
@@ -418,9 +418,7 @@ func (t *http2Client) CloseStream(s *Stream, err error) {
 	delete(t.activeStreams, s.id)
 	if t.state == draining && len(t.activeStreams) == 0 {
 		// The transport is draining and s is the last live stream on t.
-		t.mu.Unlock()
-		t.Close()
-		return
+		defer t.Close()
 	}
 	t.mu.Unlock()
 	if updateStreams {

--- a/transport/http2_client.go
+++ b/transport/http2_client.go
@@ -856,11 +856,7 @@ func (t *http2Client) reader() {
 	// Check the validity of server preface.
 	frame, err := t.framer.readFrame()
 	if err != nil {
-		if t.state == draining {
-			t.Close()
-		} else {
-			t.notifyError(err)
-		}
+		t.notifyError(err)
 		return
 	}
 	sf, ok := frame.(*http2.SettingsFrame)
@@ -888,12 +884,7 @@ func (t *http2Client) reader() {
 				continue
 			} else {
 				// Transport error.
-				if t.state == draining {
-					// A network error happened after the connection is drained. Fail the connection immediately.
-					t.Close()
-				} else {
-					t.notifyError(err)
-				}
+				t.notifyError(err)
 				return
 			}
 		}
@@ -1000,11 +991,16 @@ func (t *http2Client) GoAway() <-chan struct{} {
 
 func (t *http2Client) notifyError(err error) {
 	t.mu.Lock()
-	defer t.mu.Unlock()
 	// make sure t.errorChan is closed only once.
-	if t.state == reachable || t.state == draining {
+	if t.state == draining {
+		t.mu.Unlock()
+		t.Close()
+		return
+	}
+	if t.state == reachable {
 		t.state = unreachable
 		close(t.errorChan)
 		grpclog.Printf("transport: http2Client.notifyError got notified that the client transport was broken %v.", err)
 	}
+	t.mu.Unlock()
 }

--- a/transport/http2_client.go
+++ b/transport/http2_client.go
@@ -454,7 +454,7 @@ func (t *http2Client) Close() (err error) {
 		t.mu.Unlock()
 		return
 	}
-	if t.state == reachable {
+	if t.state == reachable || t.state == draining {
 		close(t.errorChan)
 	}
 	t.state = closing
@@ -856,7 +856,11 @@ func (t *http2Client) reader() {
 	// Check the validity of server preface.
 	frame, err := t.framer.readFrame()
 	if err != nil {
-		t.notifyError(err)
+		if t.state == draining {
+			t.Close()
+		} else {
+			t.notifyError(err)
+		}
 		return
 	}
 	sf, ok := frame.(*http2.SettingsFrame)
@@ -884,7 +888,12 @@ func (t *http2Client) reader() {
 				continue
 			} else {
 				// Transport error.
-				t.notifyError(err)
+				if t.state == draining {
+					// A network error happened after the connection is drained. Fail the connection immediately.
+					t.Close()
+				} else {
+					t.notifyError(err)
+				}
 				return
 			}
 		}
@@ -993,7 +1002,7 @@ func (t *http2Client) notifyError(err error) {
 	t.mu.Lock()
 	defer t.mu.Unlock()
 	// make sure t.errorChan is closed only once.
-	if t.state == reachable {
+	if t.state == reachable || t.state == draining {
 		t.state = unreachable
 		close(t.errorChan)
 		grpclog.Printf("transport: http2Client.notifyError got notified that the client transport was broken %v.", err)

--- a/transport/http2_server.go
+++ b/transport/http2_server.go
@@ -680,6 +680,11 @@ func (t *http2Server) controller() {
 					t.framer.writeRSTStream(true, i.streamID, i.code)
 				case *goAway:
 					t.mu.Lock()
+					if t.state == closing {
+						t.mu.Unlock()
+						// The transport is closing.
+						return
+					}
 					sid := t.maxStreamID
 					t.state = draining
 					t.mu.Unlock()

